File: lib/core_services/service_init.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'interfaces/export.dart';

part 'service_init.freezed.dart';
part 'service_init.g.dart';

@freezed
sealed class ServiceInit with _$ServiceInit {
  const factory ServiceInit({
    required String name,
    String? uid,
    required OImage faviconUrl,
    required String rootUrl,
    String? version,
    String? description,
    /// Code language support
    String? language,
    List<SettingField>? settings,
    List<WebRule>? webRules,
    @Default(false) bool fetchHeadless,
    String? fetchBaseUrl,

    ///
    /// {BASE_URL} = baseUrl. Example "{BASE_URL}/captcha" = https://example.com/captcha
    ///
    String? captchaUrl,

    ///
    /// {OLD_COOKIE} = old cookie
    ///
    String? customCookie,
  }) = _ServiceInit;

  factory ServiceInit.fromJson(Map<String, dynamic> json) =>
      _$ServiceInitFromJson(json);
}

File: lib/core_services/service_init.ts
import type { OImage } from "./interfaces/o_image"
import type { SettingField } from "./interfaces/setting/setting_field"
import type { WebRule } from "./interfaces/web_rule"

export interface ServiceInit {
  name: string
  uid?: string
  faviconUrl: OImage
  rootUrl: string
  version?: string
  description?: string
  language?: string
  captchaUrl?: () => string
  settings?: SettingField[]
  webRules?: WebRule[]
  fetchHeadless?: boolean
  fetchBaseUrl?: string
  onBeforeInsertCookie?: (oldCookie?: string) => string | undefined
}
